### ISSUE TITLE
Inserting a bad HTML code in the code editor causes sometimes (too many times ...) further HTML tags to disappear ...

### ISSUE THREAD

**Issue by ghost** [maqetta/maqetta#3521]
Example: I put an <img> tag in the sources and edited it by adding src="image/file.png". Now the image shows well but the rest of the HTML has vanished. Forever ........ 
Please consider adding some sort of backup mechanism until all such bugs are fixed.

I just got my day's work ruined

**Comment by JonFerraiolo** [comment]
Sorry about losing your work. Definitely high priority.

**Comment by JonFerraiolo** [comment]
@peller @jhpedemonte 

I played around with things and I have a **THEORY** about what might be happening in this case. 

If a user is in split view, and they have an `<img>` element is near the top of the document (e.g., a banner image), and the user partly types the following into the source pane:

```
<img src='aa.png></img>
```

of if the user types in the incorrect close quote (e.g., uses single-quote to start an attribute by double-quote to end the end attribute), after pausing for a second (so Maqetta is no longer waiting to process a batch of keystrokes),  the Maqetta HTML parser runs on the content and assumes the whole rest of the document represents the value of the 'src' attribute. The design canvas will become empty (i.e., only the `<img>` element will show). It looks to the user as if the whole document has been wiped out.

However, the source code is all still there. If the user finishes with his edits so that the document is well-formed again (i.e., by adding the proper closing quote):

```
<img src='aa.png'></img>
```

then the previously invisible parts of the document reappear.

However, this is just a theory. @rubimazaki if you are able to provide more details, such as the complete source contents when the screen goes blank, that would be helpful in moving from theory to fact.

Regardless, we definitely need a "go back in history" feature so that when users see things go amiss with their composition, they can easily go back to a previously working version of their content. (Although note that such a timeline feature may be non-trivial to implement)

**Comment by ghost** [comment]
Hi,

I can contribute that I was using multiple nested BorderContainer controls.
Seems it's the best way to layout controls in Maqetta.
But anyway, I think the fundamental issue here, is if the editor does not
understand anything, the best fallback is just to leave previous code
as-is. Or better mark it as an error.

Current behavior is really bad in my opinion. _Maqetta should not attempt
to read and alter code written in html and css files_ (had another issue
with a css file written in somewhat different style, maqetta chaged and it
was rendered useless ... I had a backup than...). It would be best if upon
error the designer would say: "No! first fix the bad syntax issues, than
come back and design". So poorly written code (like unclosed " or tag> )
will not cause huge chunks of code to be removed as result of it being
regarded as junk (which seems to be the current behavior...)

Thanks

**Comment by JonFerraiolo** [comment]
@rubimazaki I agree with what you are suggesting, and I agree that this problem of problems is one of worst shortcomings with Maqetta, and I agree with all of the reasons you have provided.

**Comment by peller** [comment]
see #2268

**Comment by peller** [comment]
isn't this only a problem if the user hits 'save'?  It might be good to restore the "white screen of death" we had in #2268 where the Dojo parser fails, and perhaps even restrict auto saves until that situation is corrected, but Jon provides a good example where incomplete HTML syntax can break the page, which has nothing to do with the parser, and we're not doing any HTML validation on the document (not sure we could ever fully trust such a tool anyway, if we did)

**Comment by ghost** [comment]
(Sorry for putting this in a list //:)
1. not only on save. Simply going from source->design does that, and of course there's no UNDO .. only a button
2. I have found another bug that messed up my workspace: using < br > is allowed in HTML, but maqetta expects < /br > right after it. If it doesn't find, the trouble begins.. It changes the order of the next tabs in a crazy way, but you just might not notice that immediately, only when it's too late.
3. A solid undo buffer that incorporates the source editing would be best. I know it's hard, but it might be a good step to start with.
4. I don't agree that a HTML syntax validator is such a hard thing to do. It's not C++ .. just verify that the correct tags are closed (and comply with HTML rules).
5.  Border controllers are the only (at least for a non HTML guru) way of making layouts in Maqetta. They are powerful but also buggy (editing them). there are many issues using them, especially when used nested.
6. Please consider adding layout controls (BorderControls with their padding set to 0 and without gutters are a good start..)

Please also consider taking some time to make Maqetta more stable. It's a great tool, but with the issues it currently holds it's just impossible to make a sane work. 

Thanks

**Comment by JonFerraiolo** [comment]
@rubimazaki - Thanks for your latest great suggestions and feedback. We will do our best to address these things as soon as we can.

**Comment by JonFerraiolo** [comment]
Unfortunately, ran out of time on this one for M8 release. Pushing to M9.

**Comment by JonFerraiolo** [comment]
See https://github.com/maqetta/maqetta/commit/6108f37f3d0d1f5f2be884c554ffd300759bfd1f#commitcomment-2462855

**Comment by JonFerraiolo** [comment]
For M10, things are a little better - we will be adding undo support for source edits. However, the application can still get in a bad state after source editing, and recovery is sometimes not possible. To go further on this issue will require some major investment (e.g., update the parsers, add cross-session undos) that won't happen in the M10 timeframe, so pushing the remainder to M11.
